## Token Embedding
Now that we understand tokenization, we have solved one part of the problem of representing language to a language model. In this sense, language is a sequence of tokens. And if we train a good-enough model on a large-enough set of tokens, it starts to capture the complex patterns that appear in its training dataset:
 - If the training data contains a lot of English text, that pattern reveals itself as a model capable of representing and generating the English language
 - If the training data contains factual information (Wikipedia, for example), the model would have the ability to generate some factual information (see the following note).

### A Language Model Holds Embeddings for the Vocabulary of Its Tokenizer

- After a tokenizer is initialized and trained, it is then used in the training process of its associated language model.

- This is why a pretrained language model is linked with its tokenizer and can’t use a different tokenizer without training.
- The language model holds an embedding vector for each token in the tokenizer’s vocabulary
- Before the beginning of the training process, these vectors are randomly initialized like the rest of the model’s weights, but the training process assigns them the values that enable the useful behavior they’re trained to perform.



### Creating Contextualized Word Embeddings with Language Models

In [1]:
from transformers import AutoModel, AutoTokenizer

# Load the tokenizer
tokenizer = AutoTokenizer.from_pretrained("microsoft/deberta-base")

# Load a language model
model = AutoModel.from_pretrained("microsoft/deberta-v3-xsmall")


# Tokenize the sentence
tokens = tokenizer('Hello world', return_tensors='pt')

# Process the tokens
output = model(**tokens)[0]

d:\2026-courses\LLMs-Handson\venv\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\pc\.cache\huggingface\hub\models--microsoft--deberta-base. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)
d:\2026-courses\LLMs-Handson\venv\Lib\site-packages\huggingface_hub\file_download.py:942: FutureWarning: `resume_do

In [2]:
output.shape

torch.Size([1, 4, 384])

- If we skip the first dimension, we can read this as 4 tokens, each token is embedded in a vector of 384 values.
- The first dimension is the batch dimension used in cases (like training) when we want to send multiple input sentences to the model at the same time (they’re processed at the same time, which speeds up the process).

In [3]:
for token  in tokens['input_ids'][0]:
    print(tokenizer.decode([token]))

[CLS]
Hello
 world
[SEP]


- This particular tokenizer and model operate by adding the [CLS] and [SEP] tokens to the beginning and end of a string.
- Our language model has now processed the text input. The result of its output is the following:

In [4]:
output

tensor([[[-3.4816,  0.0861, -0.1819,  ..., -0.0612, -0.3911,  0.3017],
         [ 0.1898,  0.3208, -0.2315,  ...,  0.3714,  0.2478,  0.8048],
         [ 0.2071,  0.5036, -0.0485,  ...,  1.2175, -0.2292,  0.8582],
         [-3.4278,  0.0645, -0.1427,  ...,  0.0658, -0.4367,  0.3834]]],
       grad_fn=<NativeLayerNormBackward0>)